File: Ejercicios/T4/CODIGO/PYTHON/THough/hough_lines.py
import cv2
import numpy as np
import sys

def onTrackbarChange(max_slider):
	global img
	global dst
	global gray

	dst = np.copy(img)

	th1 = max_slider 
	th2 = th1 * 0.4
	edges = cv2.Canny(img, th1, th2)
	
	# Tranformada de Houg probabilistica para rectas
	lines = cv2.HoughLinesP(edges, 2, np.pi/180.0, 50, minLineLength=10, maxLineGap=100)

	# Debuxamos as linhas
	for line in lines:
		x1, y1, x2, y2 = line[0]
		cv2.line(dst, (x1, y1), (x2, y2), (0,0,255), 1)

	cv2.imshow("Imaxe Resultado", dst)	
	cv2.imshow("Bordes",edges)

if __name__ == "__main__":
	
	# Lemos a imaxe
	img = cv2.imread('./data/lanes.jpg')
	if img is None:
		print("Non atopo a imaxe")
	
	# Copia para usos posteriores
	dst = np.copy(img)

	# Convertemos a gros
	gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

	# Creamos xanelas de bisualizacion
	cv2.namedWindow("Bordes")
	cv2.namedWindow("Imaxe Resultado")
	  

	# Inicializamos o limiar
	initThresh = 500

	# Maximo valor do limiar
	maxThresh = 1000

	cv2.createTrackbar("Limiar", "Imaxe Resultado", initThresh, maxThresh, onTrackbarChange)
	onTrackbarChange(initThresh)

	while True:
		key = cv2.waitKey(1)
		if key == 27:   #Pulsa Esc
			break

	cv2.destroyAllWindows()

File: Ejercicios/T4/CODIGO/C++/THough/hough_lines.cpp
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/opencv.hpp>
#include <stdio.h>
#include <iostream>

using namespace cv;
using namespace std;

// Variables para almacenar imaxes
Mat dst, cimg, gray, img, edges;

int initThresh;
const int maxThresh = 1000;
double th1,th2;

// vector para almacenar os puntos das linhas
vector<Vec4i> lines;

void onTrackbarChange( int , void* )
{ 
  dst = img.clone();

  th1 = initThresh;
  th2 = th1 * 0.4;

  Canny(img,edges,th1,th2);
  
  // Detectamos linhas empregando a transforamda de HoughLinesP (probabilistica)
  HoughLinesP(edges, lines, 2, CV_PI/180, 50, 10, 100);

  // Debuxamos as linhas sobre os puntos detectados
   for( size_t i = 0; i < lines.size(); i++ )
    {
        Vec4i l = lines[i];
        line( dst, Point(l[0], l[1]), Point(l[2], l[3]), Scalar(0,0,255), 1, LINE_AA);
    }
   
   // Visualizamos os resultados
   imshow("Imaxe resultado", dst);
   imshow("bordes", edges);
}

int main(int argc, char** argv) {
  const char* file = argv[1];
  // Read image (color mode)
  img = imread(file, 1);
  dst = img.clone();

  if(img.empty())
  {
    cout << "Erro lendo a imaxe" << file<< endl;
    return -1;
  }

  // Convertemos a gris
  cvtColor(img, gray, COLOR_BGR2GRAY);

  // Detectamos borde mediane Canny
  // Canny(gray, dst, 50, 200, 3);
  
  // Facemos unha copia da imaxe orixinal
  // cimg = img.clone();

  // A efectos de visualizacion
  namedWindow("bordes",1);
  namedWindow("Imaxe resultado", 1);
 
  // Declaramos un limiar para a transformada de hough
  initThresh = 500;

  // Creamos unha barra deslizante para cambiar dinamicamente o valor do limiar
  createTrackbar("Limiar", "Imaxe resultado", &initThresh, maxThresh, onTrackbarChange);
  onTrackbarChange(initThresh, 0);

  while(true)
  {
    int key;
    key = waitKey( 1 );
    if( (char)key == 27 )
      { break; }
  }

  destroyAllWindows();
}
